thrust/system/cuda/detail/bulk/malloc.hpp

```diff
@@ -38,9 +38,24 @@ inline __device__ bool is_on_chip(void *ptr)
 template<typename T>
 inline __device__ T *on_chip_cast(T *ptr)
 {
+#if defined(__CUDA__) && defined(__NVCC__) && !defined(__clang__)
```

Comment:
nvcc doesn't define `__CUDA__` afaik, so this path won't be taken and `nvcc` will fall back to `#else`. We need something like that

    #ifdef __NVCC__  

 or

    #if !(defined(__CUDA__) && defined(__clang__))

but not 

    #if defined(__CUDA__) && defined(__NVCC__) && !defined(__clang__)

If clang is a host compiler both `__NVCC__` and `__clang__` will be defined.